airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/thread_based_concurrent_stream.py

```diff
@@ -135,10 +142,18 @@ def _wait_while_too_many_pending_futures(self, futures: List[Future[Any]]) -> No
     def _check_for_errors(self, futures: List[Future[Any]]) -> None:
         exceptions_from_futures = [f for f in [future.exception() for future in futures] if f is not None]
         if exceptions_from_futures:
-            raise RuntimeError(f"Failed reading from stream {self.name} with errors: {exceptions_from_futures}")
-        futures_not_done = [f for f in futures if not f.done()]
-        if futures_not_done:
-            raise RuntimeError(f"Failed reading from stream {self.name} with futures not done: {futures_not_done}")
+            exception = RuntimeError(f"Failed reading from stream {self.name} with errors: {exceptions_from_futures}")
+            self._stop_and_raise_exception(exception)
+        else:
+            futures_not_done = [f for f in futures if not f.done()]
+            if futures_not_done:
+                exception = RuntimeError(f"Failed reading from stream {self.name} with futures not done: {futures_not_done}")
+                self._stop_and_raise_exception(exception)
+
+    def _stop_and_raise_exception(self, exception: BaseException) -> None:
```

Comment:
This is interesting. I had in mind that we would read from all the partitions anyway and the state would reflect that. This is only true for incremental syncs though and there is no need to continue processing for full refresh syncs.

In any case, I think I approve of this new behavior. I just wanted to share my previous reasoning and if it rang a bell at any level.

Reply:
I'm generally in favor of failing early because it's the least surprising behavior, but I see your point about being able to sync other partitions for incremental syncs. I think that's a good follow up optimization for incremental reads

Reply:
Ok! Let's wait to see if there is value in this. I suspect the value is not so high